### PR TITLE
Preferred portal refresh

### PR DESCRIPTION
**Pull request by m-cat**
# PULL REQUEST

## Overview

- Implements redirecting the skapp to be on the user's preferred portal, if he has a preferred portal configured.
- Updates some URL utilities and tests based on edge cases that were found when implementing the redirect.

**Note:** Has not been tested yet. Waiting for the corresponding MySky changes, then will test on `alpha` MySky.

## Full flow

```org
 Load MySky:
 1. SDK opens MySky on the same portal as the skapp.
 2. MySky always connects to siasky.net first.
 3. MySky tries to get the saved portal preference.
    1. If the portal is set, MySky switches to using the preferred portal.
    2. If it is not set or we don't have the seed, MySky switches to using
       the current portal as opposed to siasky.net.
 4. After MySky finishes loading, SDK queries `mySky.getPortalPreference`.
 5. If the preferred portal is set and different than the current portal,
    SDK triggers refresh.
 6. We go back to step 1 and repeat, but since we're on the right portal now
    we won't refresh in step 5.

Login:
1. SDK logs in either silently or through the UI.
2. If it was through the UI, MySky switches to siasky.net and tries to get
   the saved portal preference.
   1. If the portal is set, MySky switches to using the preferred portal.
   2. If it is not set or we don't have the seed, MySky switches to using
      the current portal as opposed to siasky.net.
3. SDK queries `mySky.getPortalPreference`.
4. If the preferred portal is set and different than the current portal, SDK
   triggers refresh.
5. We go to "Load MySky" step 1 and go through that flow, but we don't
   refresh in step 5.
```

## Checklist
Review and complete the checklist to ensure that the PR is complete before assigned to an approver.
 - [ ] All new methods or updated methods have clear docstrings
 - [ ] Testing added or updated for new methods
 - [ ] Verify if any changes impact the WebPortal Health Checks
 - [ ] Approriate documentation updated
 - [ ] Changelog file created

### CODE REVIEW

File: src/mysky/index.ts [hunk 1]
```diff
@@ -765,6 +797,10 @@ export class MySky {
     return connection;
   }
 
+  protected async getPreferredPortal(): Promise<string | null> {
```
Comment (peterjan) [inline src/mysky/index.ts:800]: missing docstring

File: src/mysky/index.ts [hunk 2]
```diff
@@ -774,18 +810,76 @@ export class MySky {
...
-            // Don't throw on error, just print a console warning.
-            console.warn(error);
-          }
-        })
+    if (!loggedIn) {
+      return;
+    }
+
+    // Call the `onUserLogin` hook for all DACs.
+    await Promise.all(
```
Comment (peterjan) [inline src/mysky/index.ts:824]: I don't think you want to use `Promise.all` here. If one DAC immediately rejects it could make it so `onUserLogin` is not called on other DACs. I think we want to use `.allSettled` here.

Comment (peterjan) [reply src/mysky/index.ts:824]: Oh lol never mind, I just saw you wrap the call in a `try-catch` anyway. I'll leave the comment in the off-chance you were unaware of the behavior of `Promise.all`, but feel free to resolve it.

File: src/mysky/utils.ts [hunk 3]
```diff
@@ -15,6 +16,26 @@ export async function getFullDomainUrl(this: SkynetClient, domain: string): Prom
   return getFullDomainUrlForPortal(portalUrl, domain);
 }
 
+// TODO: unit test
```
Comment (peterjan) [inline src/mysky/utils.ts:19]: probably not something we want to follow up with but rather implement in this PR?

File: src/mysky/utils.ts [hunk 4]
```diff
@@ -64,3 +85,20 @@ export function popupCenter(url: string, winName: string, w: number, h: number):
...
+ * Returns whether we should redirect from the current portal to the preferred
+ * portal. The protocol prefixes are allowed to be different and there can be
+ * other differences like a trailing slash.
+ *
+ * @param currentPortalUrl - The current portal URL.
+ * @param preferredPortalUrl - The preferred portal URL.
+ * @returns - Whether the two URLs are equal for the purposes of redirecting.
+ */
+export function shouldRedirectToPreferredPortalUrl(currentPortalUrl: string, preferredPortalUrl: string): boolean {
+  currentPortalUrl = currentPortalUrl.split("//", 2)[1] || currentPortalUrl;
```
Comment (peterjan) [inline src/mysky/utils.ts:100]: `nit` I don't like the `split` on "//", it's just less correct even though the input is probably already sanitized in one way or the other
```
// strip protocol and trailing slash
currentPortalUrl = trimSuffix(currentPortalUrl.replace(/https:\/\/|http:\/\//, ''), "/");
preferredPortalUrl = trimSuffix(preferredPortalUrl.replace(/https:\/\/|http:\/\//, ''), "/");
return currentPortalUrl === preferredPortalUrl;
```

**Review by peterjan — DISMISSED** [review]

File: src/mysky/index.ts (hunk 1 above)
Comment (m-cat) [reply src/mysky/index.ts:800]: Looks like `require-jsdoc` does not require method documentation by default. 🤯

https://eslint.org/docs/rules/require-jsdoc#options

I'll enable stricter checking for this rule.

File: src/mysky/utils.ts (hunk 3 above)
Comment (m-cat) [reply src/mysky/utils.ts:19]: The unit tests have been implemented, will remove the comment.

File: src/mysky/utils.ts (hunk 4 above)
Comment (m-cat) [reply src/mysky/utils.ts:100]: Thanks, just needed a slight modification to make it case-insensitive.

File: tsconfig.json [hunk 5]
```diff
@@ -8,6 +8,8 @@
     "noImplicitAny": true,
     "declaration": true,
     "isolatedModules": true,
+    "moduleResolution": "node",
+    "target": "es2019",
```
Comment (m-cat) [inline tsconfig.json:12]: Needed for `flatMap()` and `flat()`. Note that these `tsconfig` fields were already set in `tsconfig.build.json`.

File: utils/testing.ts [hunk 6]
```diff
@@ -3,6 +3,32 @@
...
+ * @returns - Array of arrays of all possible permutations.
+ * @see {@link https://gist.github.com/ssippe/1f92625532eef28be6974f898efb23ef#gistcomment-3530882}
+ */
+export function combineArrays<T>(...arrays: Array<Array<T>>): Array<Array<T>> {
+  return arrays.reduce<T[][]>(
+    (accArrays, array) => accArrays.flatMap((accArray) => array.map((value) => [...accArray, value])),
+    [[]]
+  );
+}
+
```
Comment (m-cat) [inline utils/testing.ts:31]: Added to simplify generating a lot of test cases and expected results.

**Review by ro-tex — DISMISSED** [review]

**Review by ro-tex — DISMISSED** [review]

File: src/mysky/index.ts [hunk 7]
```diff
@@ -161,7 +204,13 @@ export class MySky {
       permissions.push(perm1, perm2, perm3);
     }
 
-    MySky.instance = new MySky(connector, permissions, hostDomain);
+    const currentPortalUrl = await currentUrlClient.portalUrl();
+    MySky.instance = new MySky(connector, permissions, hostDomain, currentPortalUrl);
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky/index.ts:208]: > Is is possible to call this new more than once?

Nope, we enforce this above:

```ts
// Enforce singleton.
if (MySky.instance) {
  return MySky.instance;
}
```

> implement a MySky.getInstance() method which does that and is the recommended way of getting the mysky instance. It calls the constructor if there is no singleton yet or returns the singleton otherwise.

I don't see the advantage of this over the existing entry point `loadMySky`. It seems to me that `getInstance` would just be a wrapper over `loadMySky`. Let me know if I'm missing something though, I'm not familiar with the best practices around singletons.

[the remark this answers is not in the document]
Comment (peterjan) [reply src/mysky/index.ts:208]: For what it's worth, singletons that accept parameters in the constructor aren't really singletons in the first place. That's why you usually have a static `getInstance()` as ivo says, and actually you usually also make the constructor private. Only ever exposing `getInstance()` which creates the singleton if that hasn't been done yet.

What's the reason you are calling this a singleton in this case?
The way it's created is a bit weird if I'm honest, the `instance` is not 'set' on the class from the outside, it should even be private and get returned by `getInstance()`.

I understand what you're doing though, I'm not sure I'd call it a singleton. If possible I would even drop the `instance` and simply work with a global variable perhaps. You can resolve this comment though if you don't care about all of this and want to keep it the way it is. This is all coming from textbooks so I honestly don't care that much about it. But for someone that does care it will definitely look a bit odd.

[the remark this answers is not in the document]
Comment (peterjan) [reply src/mysky/index.ts:208]:
```
class MySingleton {
     private static instance: MySingleton;
     public static getInstance(): MySingleton {
          if (!this.instance) { 
              this.instance = new MySingleton();
          }
          return this.instance;
     }

     private constructor() {}
}

[the remark this answers is not in the document]
Comment (peterjan) [reply src/mysky/index.ts:208]: The more I think about it the more I think we should refactor it slightly. Especially the following:
```
if (MySky.instance) {
      return MySky.instance;
}
```
is actually quite dangerous and misleading because it totally neglects the parameters with which the function was called, right?

[the remark this answers is not in the document]
Comment (peterjan) [reply src/mysky/index.ts:208]: I'm also not really a fan of the `Error` thrown in the MySky constructor. I would make it possible for multiple MySky instances to be created, but structure the code in a way that only one is ever created. That's more easily said than done of course but perhaps in this case the easiest thing to do is use a global variable.

What's the reason only one MySky instance should get created by the way? Perhaps there's no real good reason and we don't have to care about all of this.

[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky/index.ts:208]: Created https://github.com/SkynetLabs/skynet-js/issues/374 so I can revisit the singleton pattern and whether it's even needed at a more convenient time, probably involving Karol and Daniel in the discussion.

**Review by ro-tex — COMMENTED** [review]
Aside from the MySky singleton issue, it looks fine. Let's resolve that one and merge this.

File: src/client.test.ts [hunk 8]
```diff
@@ -110,3 +110,70 @@ describe("buildRequestUrl", () => {
     });
   });
 });
+
+describe("client options", () => {
+  const skylink = "XABvi7JtJbQSMAcDwnUnmp2FKDPjg8_tTTFP4BwMSxVdEg";
+  const sialink = `${URI_SKYNET_PREFIX}${skylink}`;
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/client.test.ts:116]: Maybe. We'd have to deprecate this one. It's already being exported so we can't just rename it. Not sure it's worth it.

File: src/mysky/index.ts [hunk 9]
```diff
@@ -132,10 +136,35 @@ export class MySky {
...
+    permissions: Permission[],
+    protected hostDomain: string,
+    protected currentPortalUrl: string
+  ) {
+    if (MySky.instance) {
+      throw new Error("Trying to create a second MySky instance");
+    }
+
     this.pendingPermissions = permissions;
   }
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky/index.ts:158]: I honestly don't have the bandwidth to be thinking about non-urgent things right now. This was not introduced in this PR, and in my opinion is "low in urgency, low in importance". Better to get to it later when we can have a proper think about it (and maybe Karol or Michał can provide input). If this is important to you right now and you have time then feel free to contribute a commit.

I'll save your comment in the associated issue. Thanks for writing it out - I just can't get pulled into random things right now.

> I'd prefer to solve this here rather than postponing it to a F/U which might fall through the cracks.

I am aware of the outstanding issues in the projects I am responsible for. It's not really a matter of falling through the cracks, but of being chronically overworked for at least the past year.

File: src/mysky/index.ts [hunk 10]
```diff
@@ -776,26 +908,157 @@ export class MySky {
...
-          try {
-            await dac.onUserLogin();
-          } catch (error) {
-            // Don't throw on error, just print a console warning.
-            console.warn(error);
-          }
-        })
+    if (!loggedIn) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky/index.ts:919]: I was originally thinking there would be some logic here that runs if the login failed.

File: src/client.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (ro-tex) [reply src/client.test.ts:116]: Oh, right. I didn't think about that. No, it's not worth it. Let's keep it.

File: src/mysky/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ro-tex) [reply src/mysky/index.ts:217]: As discussed under https://github.com/SkynetLabs/skynet-js/pull/361#discussion_r804861103, I now agree that moving this to a F/U is a good solution.

**Review by peterjan — APPROVED** [review]
